<a href="https://colab.research.google.com/github/shravankoninti/AV/blob/master/Innoplex_Nvidia_23_03_2019_v0_blog_0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
# memory footprint support libraries/code
!ln -sf /opt/bin/nvidia-smi /usr/bin/nvidia-smi
!pip install gputil
!pip install psutil
!pip install humanize
import psutil
import humanize
import os
import GPUtil as GPU
GPUs = GPU.getGPUs()
# XXX: only one GPU on Colab and isn’t guaranteed
gpu = GPUs[0]
def printm():
  process = psutil.Process(os.getpid())
  print("Gen RAM Free: " + humanize.naturalsize( psutil.virtual_memory().available ), " | Proc size: " + humanize.naturalsize( process.memory_info().rss))
  print("GPU RAM Free: {0:.0f}MB | Used: {1:.0f}MB | Util {2:3.0f}% | Total {3:.0f}MB".format(gpu.memoryFree, gpu.memoryUsed, gpu.memoryUtil*100, gpu.memoryTotal))
printm() 

  Stored in directory: /root/.cache/pip/wheels/3d/77/07/80562de4bb0786e5ea186911a2c831fdd0018bda69beab71fd
Successfully built gputil
Gen RAM Free: 12.9 GB  | Proc size: 142.5 MB
GPU RAM Free: 11441MB | Used: 0MB | Util   0% | Total 11441MB


In [0]:
RANDOM_STATE = 17

In [0]:
import pandas as pd
import numpy as np
from sklearn.feature_extraction import DictVectorizer
from sklearn.feature_extraction.text import HashingVectorizer
from sklearn.linear_model import Perceptron
from sklearn.model_selection import train_test_split
from sklearn.linear_model import SGDClassifier
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import classification_report
import nltk

In [4]:
# Install the PyDrive wrapper & import libraries.
# This only needs to be done once in a notebook.
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials

# Authenticate and create the PyDrive client.
# This only needs to be done once in a notebook.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

    100% |████████████████████████████████| 993kB 19.0MB/s 


In [0]:
#train.csv

# Download a file based on its file ID.
# https://drive.google.com/open?id=1A3BLkSn7GOwR7J2ccoIoDoQVOXLeytyR

file_id = '1A3BLkSn7GOwR7J2ccoIoDoQVOXLeytyR' #https://drive.google.com/open?id=1R72cIA1deLDdg0aULjcl5OR6tsv-DRaq
downloaded = drive.CreateFile({'id': file_id})
downloaded.GetContentFile('sample_submission.csv')

In [6]:
submission = pd.read_csv('sample_submission.csv', encoding = "ISO-8859-1")
submission.head()

id  Sent_ID tag
0  4543834   191283   O
1  4543835   191283   O
2  4543836   191283   O
3  4543837   191283   O
4  4543838   191283   O

In [7]:
submission.shape

(2994463, 3)

In [0]:
#test.csv

file_id = '1A74kdE3pv_WvHJwuEbeF2TlYk4irXlis' #https://drive.google.com/open?id=1R72cIA1deLDdg0aULjcl5OR6tsv-DRaq
downloaded = drive.CreateFile({'id': file_id})
downloaded.GetContentFile('test.csv')

In [9]:
test_df = pd.read_csv('test.csv', encoding = "ISO-8859-1")
test_df.head(5)

id  Doc_ID  Sent_ID    Word
0  4543834   30001   191283   CCCVA
1  4543835   30001   191283       ,
2  4543836   30001   191283  MANOVA
3  4543837   30001   191283       ,
4  4543838   30001   191283      my

In [10]:
test_df.shape

(2994463, 4)

In [0]:
#train.csv

# Download a file based on its file ID.

file_id = '1R72cIA1deLDdg0aULjcl5OR6tsv-DRaq' #https://drive.google.com/open?id=1R72cIA1deLDdg0aULjcl5OR6tsv-DRaq
downloaded = drive.CreateFile({'id': file_id})
downloaded.GetContentFile('train.csv')

In [12]:
train_df = pd.read_csv('train.csv', encoding = "ISO-8859-1")
train_df.head()

id  Doc_ID  Sent_ID           Word tag
0   1       1        1        Obesity   O
1   2       1        1             in   O
2   3       1        1           Low-   O
3   4       1        1            and   O
4   5       1        1  Middle-Income   O

In [13]:
train_df.isnull().sum()

id           0
Doc_ID       0
Sent_ID      0
Word       130
tag          0
dtype: int64

In [14]:
train_df = train_df.fillna(method='ffill')
print(train_df.Sent_ID.nunique(), train_df.Word.nunique(), train_df.tag.nunique())

test_df = test_df.fillna(method='ffill')
print(test_df.Sent_ID.nunique(), test_df.Word.nunique())

train_df['id_Sent_ID'] = train_df['id'] + train_df['Sent_ID']
test_df['id_Sent_ID'] = test_df['id'] + test_df['Sent_ID']

191282 184505 3
125840 139891


In [15]:
train_df.head(2)

id  Doc_ID  Sent_ID     Word tag  id_Sent_ID
0   1       1        1  Obesity   O           2
1   2       1        1       in   O           3

## Data Preprocessing

In [16]:
train_df.isnull().sum()

id            0
Doc_ID        0
Sent_ID       0
Word          0
tag           0
id_Sent_ID    0
dtype: int64

We have 191,282 sentences that contain 184,505 unique words and tagged by 3 tags.

The tags are not evenly distributed.

In [17]:
train_df.groupby('tag').size().reset_index(name='counts')

tag   counts
0  B-indications    53003
1  I-indications    44624
2              O  4446206

In [18]:
test_df.isnull().sum()

id            0
Doc_ID        0
Sent_ID       0
Word          0
id_Sent_ID    0
dtype: int64

In [0]:
# import nltk
# nltk.download('all')

In [20]:
import nltk
nltk.download('averaged_perceptron_tagger')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


True

In [0]:
train_df['POS_full'] = nltk.pos_tag(train_df['Word']) 
# test_df['POS_full'] = nltk.pos_tag(test_df['Word']) 

In [22]:
train_df.head(5)

id  Doc_ID  Sent_ID           Word tag  id_Sent_ID             POS_full
0   1       1        1        Obesity   O           2        (Obesity, NN)
1   2       1        1             in   O           3             (in, IN)
2   3       1        1           Low-   O           4          (Low-, NNP)
3   4       1        1            and   O           5            (and, CC)
4   5       1        1  Middle-Income   O           6  (Middle-Income, JJ)

In [104]:
test_df.head(5)

id  Doc_ID  Sent_ID    Word  id_Sent_ID       POS_full
0  4543834   30001   191283   CCCVA     4735117   (CCCVA, NNP)
1  4543835   30001   191283       ,     4735118         (,, ,)
2  4543836   30001   191283  MANOVA     4735119  (MANOVA, NNP)
3  4543837   30001   191283       ,     4735120         (,, ,)
4  4543838   30001   191283      my     4735121     (my, PRP$)

In [0]:
train_df['POS'] = train_df['POS_full'].apply(lambda x: x[1] )
test_df['POS'] = test_df['POS_full'].apply(lambda x: x[1] )

In [106]:
train_df.head(5)

id  Doc_ID  Sent_ID           Word tag  id_Sent_ID             POS_full  \
0   1       1        1        Obesity   O           2        (Obesity, NN)   
1   2       1        1             in   O           3             (in, IN)   
2   3       1        1           Low-   O           4          (Low-, NNP)   
3   4       1        1            and   O           5            (and, CC)   
4   5       1        1  Middle-Income   O           6  (Middle-Income, JJ)   

   POS  
0   NN  
1   IN  
2  NNP  
3   CC  
4   JJ

In [107]:
test_df.head(5)

id  Doc_ID  Sent_ID    Word  id_Sent_ID       POS_full   POS
0  4543834   30001   191283   CCCVA     4735117   (CCCVA, NNP)   NNP
1  4543835   30001   191283       ,     4735118         (,, ,)     ,
2  4543836   30001   191283  MANOVA     4735119  (MANOVA, NNP)   NNP
3  4543837   30001   191283       ,     4735120         (,, ,)     ,
4  4543838   30001   191283      my     4735121     (my, PRP$)  PRP$

In [0]:
# train_df = train_df.drop(['Doc_ID', 'POS_full','id'], axis= 1)
# test_df = test_df.drop(['Doc_ID', 'POS_full','id'], axis= 1)

The following code transform the text date to vector using DictVectorizer and then split to train and test sets.

In [109]:
train_df.id_Sent_ID.nunique(), train_df.Word.nunique(), train_df.POS.nunique()

(4543833, 184505, 44)

In [110]:
test_df.id_Sent_ID.nunique(), test_df.Word.nunique(), test_df.POS.nunique()

(2994463, 139891, 44)

In [111]:
test_df.isnull().sum()

Sent_ID       0
Word          0
id_Sent_ID    0
POS           0
dtype: int64

In [112]:
test_df.columns

Index(['Sent_ID', 'Word', 'id_Sent_ID', 'POS'], dtype='object')

In [0]:
test_df.to_csv('test_df_mod.csv')

In [25]:
train_df.head(3)

id  Doc_ID  Sent_ID     Word tag  id_Sent_ID       POS_full  POS
0   1       1        1  Obesity   O           2  (Obesity, NN)   NN
1   2       1        1       in   O           3       (in, IN)   IN
2   3       1        1     Low-   O           4    (Low-, NNP)  NNP

In [0]:
train_df.to_csv('train_df_mod.csv')

### Time Consuming - Hence you can start here!!

In [0]:

# file_id = '12JB8bi4VVNWfXZrWOxs5AOPgFblBiBpn'
# downloaded = drive.CreateFile({'id': file_id})
# downloaded.GetContentFile('train_df_mod.csv')


file_id = '1b75Rzb_XX32mzfoFFKkCFUZBGW419U30'
downloaded = drive.CreateFile({'id': file_id})
downloaded.GetContentFile('test_df_mod.csv')


# train_df = pd.read_csv('train_df_mod.csv')
test_df = pd.read_csv('test_df_mod.csv')

In [27]:
train_df.head(3)

id  Doc_ID  Sent_ID     Word tag  id_Sent_ID       POS_full  POS
0   1       1        1  Obesity   O           2  (Obesity, NN)   NN
1   2       1        1       in   O           3       (in, IN)   IN
2   3       1        1     Low-   O           4    (Low-, NNP)  NNP

In [46]:
train_df.tail(5)

Sent_ID                Word tag POS
4543828   191282       translational   O  JJ
4543829   191282                  or   O  CC
4543830   191282  post-translational   O  JJ
4543831   191282               level   O  NN
4543832   191282                   .   O   .

In [28]:
test_df.head(3)

Unnamed: 0  Sent_ID    Word  id_Sent_ID  POS
0           0   191283   CCCVA     4735117  NNP
1           1   191283       ,     4735118    ,
2           2   191283  MANOVA     4735119  NNP

In [29]:
train_df = train_df[['Sent_ID', 'Word', 'tag', 'POS']]

# train_df = train_df[:55000]
print(train_df.shape)


test_df = test_df[['Sent_ID', 'Word',  'POS']]
print(test_df.shape)

(4543833, 4)
(2994463, 3)


In [32]:
words = list(set(train_df["Word"].values))
n_words = len(words); n_words

184505

### We start by writing a small class to retrieve a sentence from the dataset.

In [0]:
class SentenceGetter(object):
    
    def __init__(self, data):
        self.n_sent = 1
        self.data = data
        self.empty = False
    
    def get_next(self):
#         try:
#             s = self.data[self.data["Sent_ID"] == "Sentence: {}".format(self.n_sent)]
#             self.n_sent += 1
            return self.data["Word"].values.tolist(), self.data["POS"].values.tolist(), self.data["tag"].values.tolist()    
#         except:
#             self.empty = True
#             return None, None, None

In [0]:
getter = SentenceGetter(train_df)

In [0]:
sent, pos, tag = getter.get_next()

In [64]:
### This is how a sentence looks.
print(sent); print(pos); print(tag)

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [0]:
from sklearn.base import BaseEstimator, TransformerMixin


class MemoryTagger(BaseEstimator, TransformerMixin):
    
    def fit(self, X, y):
        '''
        Expects a list of words as X and a list of tags as y.
        '''
        voc = {}
        self.tags = []
        for x, t in zip(X, y):
            if t not in self.tags:
                self.tags.append(t)
            if x in voc:
                if t in voc[x]:
                    voc[x][t] += 1
                else:
                    voc[x][t] = 1
            else:
                voc[x] = {t: 1}
        self.memory = {}
        for k, d in voc.items():
            self.memory[k] = max(d, key=d.get)
    
    def predict(self, X, y=None):
        '''
        Predict the the tag from memory. If word is unknown, predict 'O'.
        '''
        return [self.memory.get(x, 'O') for x in X]

In [66]:
tagger = MemoryTagger()
tagger.fit(sent, tag)
print(tagger.predict(sent))
tagger.tags

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



['O', 'B-indications', 'I-indications']

In [70]:
from sklearn.model_selection import train_test_split, cross_val_predict
from sklearn.metrics import classification_report


words = train_df["Word"].values.tolist()
tags = train_df["tag"].values.tolist()


pred = cross_val_predict(estimator=MemoryTagger(), X=words, y=tags, cv=5)

report = classification_report(y_pred=pred, y_true=tags)
print(report)

               precision    recall  f1-score   support

B-indications       0.72      0.40      0.52     53003
I-indications       0.59      0.37      0.46     44624
            O       0.99      1.00      0.99   4446206

    micro avg       0.98      0.98      0.98   4543833
    macro avg       0.77      0.59      0.66   4543833
 weighted avg       0.98      0.98      0.98   4543833



In [0]:
class SentenceGetter_test_df(object):
    
    def __init__(self, data):
        self.n_sent = 1
        self.data = data
        self.empty = False
    
    def get_next(self):
#         try:
#             s = self.data[self.data["Sent_ID"] == "Sentence: {}".format(self.n_sent)]
#             self.n_sent += 1
            return self.data["Word"].values.tolist(), self.data["POS"].values.tolist()    
#         except:
#             self.empty = True
#             return None, None, None

In [0]:
getter_test = SentenceGetter_test_df(test_df)

In [0]:
sent, pos = getter_test.get_next()

In [74]:
print(sent); print(pos);

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [0]:
y_pred = tagger.predict(sent)

In [78]:
np.unique(y_pred, return_counts = True)

(array(['B-indications', 'I-indications', 'O'], dtype='<U13'),
 array([  19726,   18011, 2956726]))

In [79]:
submission.head()

id  Sent_ID tag
0  4543834   191283   O
1  4543835   191283   O
2  4543836   191283   O
3  4543837   191283   O
4  4543838   191283   O

In [0]:
submission['tag'] = pd.DataFrame(y_pred)

In [81]:
submission['tag'].value_counts()

O                2956726
B-indications      19726
I-indications      18011
Name: tag, dtype: int64

In [0]:
submission.to_csv('submission_baseline_model.csv', index= False)